Topic4/src/tests/ShoppingCartTest.java

```diff
@@ -0,0 +1,274 @@
+		assertEquals(offer1, items.get(0));
+		assertEquals(item3, items.get(1));
+
+		double total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 1250.00);
+
+		shopcart.removeItem(offer1);
+		assertEquals(shopcart.getItems().size(), 1);
+		items = shopcart.getItems();
+		assertEquals(item3, items.get(0));
+
+		total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 100.00);
+
+		shopcart.removeItem(item3);
+		assertEquals(shopcart.getItems().size(), 0);
+
+		total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 0.00);
+	}
+
+	@Test
+	public void addingRemovingPayment() {
+		//adding
+		PaymentStrategy payment1 = new CashStrategy();
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+		shopcart.addPayment(payment1);
+		assertEquals(shopcart.getPayment(), payment1);
+
+		PaymentStrategy payment2 = new PaypalStrategy("luciano@gmail.com", "bootcamp");
+		shopcart.addPayment(payment2);
+		assertEquals(shopcart.getPayment(), payment2);
+		
+		//removing
+		shopcart.removePayment();
+		assertEquals(shopcart.getPayment(), null);
+	}
+	
+	@Test
+	public void payingByCreditCard() {
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+		// offer
+		Item item1 = new Product("Smart TV LG 40 inches", "TV100", 700.00);
+		Item item2 = new Product("TV 3D LG 42 inches", "TV105", 950.00);
+		Offer offer1 = new Offer("Combo TV","OF136",1400.00);
+		offer1.addItem(item1);
+		offer1.addItem(item2);
+		shopcart.addItem(offer1);
+		
+		PaymentStrategy payment = new CreditCardStrategy("Luciano",12345);
+		shopcart.addPayment(payment);
+		
+		Transaction transaction = new WebTransaction(shopcart.getItems(),payment);
+		transaction.execute();
+		
+		//paying by credit card you have a discount of 10%, the total should be 1260.00
+		assertTrue(transaction.getValue() == 1260.00);
+		
+		//the transaction has to be the number 1
+		assertEquals(1,transaction.getID());	
+	}
+	
+	@Test
+	public void payingByPaypal() {
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+
+		// offer1
+		Item item1 = new Product("Smart TV LG 40 inches", "TV100", 700.00);
+		Item item2 = new Product("Home Theatre Sony Muteki 5.1", "HT200", 600.00);
+		Offer offer1 = new Offer("Cinema in your living", "OF001", 1150.00);
+		offer1.addItem(item1);
+		offer1.addItem(item2);
+		shopcart.addItem(offer1);
+
+		// product1
+		Item item3 = new Product("Table TV O", "TT001", 100.00);
+		shopcart.addItem(item3);
+		
+		PaymentStrategy payment = new PaypalStrategy("aguirre.luciano@hotmail.com","bootcamp");
+		shopcart.addPayment(payment);
+		
+		Transaction transaction = new WebTransaction(shopcart.getItems(),payment);
+		transaction.execute();
+		
+		//paying by paypal the cheapest item is free, the total should be 1150.00
+		assertTrue(transaction.getValue() == 1150.00);
+		
+		//the transaction has to be the number 2
+		assertEquals(2,transaction.getID());	

```

Comment:
One comment to be considered: each TDD scenario should be independent of previous or later tests, in consequence to have a transaction.id equal to 2 this test should have two web transactions